src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/stormwind_city.cpp

```diff
@@ -113,28 +136,93 @@ struct npc_dashel_stonefistAI : public CombatAI
     {
         AddTimerlessCombatAction(DASHEL_LOW_HP, true);
         AddCustomAction(DASHEL_START_EVENT, true, [&]() { HandleStartEvent(); });
-        AddCustomAction(DASHEL_END_EVENT, true, [&]() { HandleEndEvent(); });
+        AddCustomAction(DASHEL_EVADE_EVENT, true, [&]() { HandleStartEvadeEvent(); }, TIMER_COMBAT_OOC);
+        AddCustomAction(DASHEL_END_EVENT, true, [&]() { HandleEndEvent(); }, TIMER_COMBAT_OOC);
     }
 
+    // old town thugs
+    ObjectGuid m_thugs[2];
+    bool m_thugsAlive;
+    // player guid to trigger: quest completed
     ObjectGuid m_playerGuid;
 
+    // EndEvent stages
+    uint32 m_phaseThugEventStage;
+
+    // check if an event has been started.
+    bool m_dialogStarted;
+
+    // used to check if its a quest fight or not.
+    bool m_questFightStarted;
+
     void Reset() override
-   {
+    {
         CombatAI::Reset();
+        if (m_questFightStarted)
+        {
+            // Reset() during quest fight -> quest failed.
+            Player* player = m_creature->GetMap()->GetPlayer(m_playerGuid);
+            if (player)
+                player->FailQuestForGroup(QUEST_MISSING_DIPLO_PT8);
+            // remove thugs
+            for (auto& guid : m_thugs)
+            {
+                if (Creature* pThug = m_creature->GetMap()->GetCreature(guid))
+                    pThug->ForcedDespawn();
+            }
+        }
+
+        // clear thug guids
+        for (auto& guid : m_thugs)
+            guid.Clear();
+
+        m_questFightStarted = false;
+        m_phaseThugEventStage = 0;
+        m_dialogStarted = false;
+        m_thugsAlive = false;
+        // restore some flags
+        m_creature->SetFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_QUESTGIVER);
+        // reset player guid
+        m_playerGuid.Clear();
+
         SetReactState(REACT_PASSIVE);
     }
 
     void JustRespawned() override
     {
         CombatAI::Reset();
-        m_creature->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PLAYER);
-        m_creature->SetFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_QUESTGIVER);
     }
 
     void StartEvent(ObjectGuid guid)
     {
+        // 2 seconds delay after quest accept npc starts attacking players.
         m_playerGuid = guid;
-        ResetTimer(DASHEL_START_EVENT, 3000);
+        ResetTimer(DASHEL_START_EVENT, 2000);
+        m_questFightStarted = true;
+    }
+
+    void HandleStartEvadeEvent()
+    {
+        // Move TargetHome after Emote
+        m_creature->GetMotionMaster()->MoveTargetedHome();
+        DisableTimer(DASHEL_EVADE_EVENT);        
```

Comment:
This isnt necessary. Custom timers dont repeat unless you explicitly tell them to.